File: server/src/personality.ts
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import * as yaml from "js-yaml";
import { generateDefaultFiles } from "./defaults.js";
import type {
  PersonaFrontmatter,
  PersonaFile,
  MemoryEntry,
  FileStatus,
  PersonaStatus,
} from "./types.js";

const HOME_PATH = path.join(os.homedir(), ".ai-personality", "personality");
export const PERSONALITY_DIR = process.env.AI_PERSONALITY_DIR
  ? path.resolve(process.env.AI_PERSONALITY_DIR)
  : HOME_PATH;

const FILE_NAMES = [
  "identity.md",
  "traits.md",
  "values.md",
  "rules.md",
  "memories.md",
  "relationships.md",
];

let _initDone = false;

export function ensurePersonalityDir(): void {
  if (_initDone) return;
  _initDone = true;
  if (!fs.existsSync(PERSONALITY_DIR)) {
    fs.mkdirSync(PERSONALITY_DIR, { recursive: true });
    console.error(`Created personality directory: ${PERSONALITY_DIR}`);
  }
  const defaults = generateDefaultFiles(PERSONALITY_DIR);
  for (const f of defaults) {
    const fp = path.join(PERSONALITY_DIR, f.filename);
    if (!fs.existsSync(fp)) {
      fs.writeFileSync(fp, f.content, "utf-8");
      console.error(`Created ${f.filename}`);
    }
  }
}

function yamlVal(v: unknown): string {
  if (Object.prototype.toString.call(v) === "[object Date]") {
    return (v as Date).toISOString().split("T")[0];
  }
  return String(v ?? "");
}

function parseFrontmatter(raw: string): { frontmatter: PersonaFrontmatter; body: string } | null {
  const match = raw.match(/^---\n([\s\S]*?)\n---\n([\s\S]*)$/);
  if (!match) return null;
  const data = yaml.load(match[1]) as Record<string, unknown>;
  return {
    frontmatter: {
      type: yamlVal(data.type),
      version: Number(data.version ?? 0),
      lastUpdated: yamlVal(data.lastUpdated),
      evolution: Number(data.evolution ?? 0),
      crossReferences: Array.isArray(data.crossReferences)
        ? (data.crossReferences as string[]).map((r) => String(r))
        : [],
    },
    body: match[2].trim(),
  };
}

function updateFrontmatterField(filename: string, field: string, value: unknown): void {
  const filePath = path.join(PERSONALITY_DIR, filename);
  const raw = fs.readFileSync(filePath, "utf-8");
  const match = raw.match(/^---\n([\s\S]*?)\n---\n([\s\S]*)$/);
  if (!match) throw new Error(`Cannot parse frontmatter in ${filename}`);
  const data = yaml.load(match[1]) as Record<string, unknown>;
  data[field] = value;
  const newFm = dumpYaml(data);
  const newRaw = `---\n${newFm}---\n${match[2]}`;
  fs.writeFileSync(filePath, newRaw, "utf-8");
}

function readFile(filename: string): PersonaFile {
  ensurePersonalityDir();
  const filePath = path.join(PERSONALITY_DIR, filename);
  const raw = fs.readFileSync(filePath, "utf-8");
  const parsed = parseFrontmatter(raw);
  return {
    filename,
    frontmatter: parsed?.frontmatter ?? {
      type: "unknown",
      version: 0,
      lastUpdated: "unknown",
      evolution: 0,
      crossReferences: [],
    },
    body: parsed?.body ?? raw,
    raw,
  };
}

export function readPersonalityFile(filename: string): PersonaFile {
  if (!FILE_NAMES.includes(filename)) {
    throw new Error(`Unknown personality file: ${filename}. Valid: ${FILE_NAMES.join(", ")}`);
  }
  return readFile(filename);
}

export function readAllPersonalityFiles(): PersonaFile[] {
  return FILE_NAMES.map((f) => readFile(f));
}

export function getSummary(): string {
  const files = readAllPersonalityFiles();
  const parts = files.map((f) => {
    const name = f.filename.replace(".md", "");
    const type = f.frontmatter.type;
    const evo = f.frontmatter.evolution;
    const refs = f.frontmatter.crossReferences.length;
    let preview = f.body.slice(0, 300);
    if (f.body.length > 300) preview += "...";
    return `=== ${name} (${type}, evolution ${evo}, ${refs} refs) ===\n${preview}`;
  });
  return parts.join("\n\n");
}

export function reflect(
  experience: string,
  lesson: string,
  affectedFiles?: string[],
): { date: string; entries: MemoryEntry[] } {
  const filePath = path.join(PERSONALITY_DIR, "memories.md");
  let raw = fs.readFileSync(filePath, "utf-8");
  const date = new Date().toISOString().split("T")[0];
  const newEntry: MemoryEntry = {
    date,
    experience,
    lesson,
    impact: "under review",
    affectedFiles: affectedFiles ?? [],
  };
  const marker = "<!-- entries -->";
  const markerIndex = raw.indexOf(marker);
  if (markerIndex === -1) {
    raw = `${raw}\n\n${marker}\n\n\`\`\`yaml\n${dumpYaml([newEntry])}\`\`\`\n`;
  } else {
    const afterMarker = raw.slice(markerIndex + marker.length);
    const yamlMatch = afterMarker.match(/```yaml\n([\s\S]*?)```/);
    if (yamlMatch) {
      const existing = cleanDates(yaml.load(yamlMatch[1])) as MemoryEntry[];
      existing.push(newEntry);
      const absStart = markerIndex + marker.length + yamlMatch.index!;
      const absEnd = absStart + yamlMatch[0].length;
      raw = raw.slice(0, absStart) + "```yaml\n" + dumpYaml(existing) + "```" + raw.slice(absEnd);
    } else {
      const insertPos = markerIndex + marker.length;
      raw = raw.slice(0, insertPos) + `\n\n\`\`\`yaml\n${dumpYaml([newEntry])}\`\`\`\n` + raw.slice(insertPos);
    }
  }
  fs.writeFileSync(filePath, raw, "utf-8");

  const entries = loadMemoryEntries(raw);
  return { date, entries };
}

function dumpYaml(data: unknown): string {
  const out = yaml.dump(data, { lineWidth: -1, noRefs: true });
  return out.replace(/['"](\d{4}-\d{2}-\d{2})['"]/g, "$1");
}

function cleanDates(v: unknown): unknown {
  if (Object.prototype.toString.call(v) === "[object Date]") return (v as Date).toISOString().split("T")[0];
  if (Array.isArray(v)) return v.map(cleanDates);
  if (v !== null && typeof v === "object") {
    const o: Record<string, unknown> = {};
    for (const [k, val] of Object.entries(v as Record<string, unknown>)) {
      o[k] = cleanDates(val);
    }
    return o;
  }
  return v;
}

function loadMemoryEntries(raw: string): MemoryEntry[] {
  const marker = "<!-- entries -->";
  const markerIndex = raw.indexOf(marker);
  if (markerIndex === -1) return [];
  const afterMarker = raw.slice(markerIndex + marker.length);
  const match = afterMarker.match(/```yaml\n([\s\S]*?)```/);
  if (!match) return [];
  return cleanDates(yaml.load(match[1])) as MemoryEntry[];
}

export function getStatus(): PersonaStatus {
  const files = readAllPersonalityFiles().map((f) => ({
    name: f.filename,
    evolution: f.frontmatter.evolution,
    lastUpdated: f.frontmatter.lastUpdated,
    refCount: f.frontmatter.crossReferences.length,
  }));

  const memFile = readFile("memories.md");
  const entries = loadMemoryEntries(memFile.raw);
  const lastReflection = entries.length > 0 ? entries[entries.length - 1].date : null;
  const pendingCount = entries.filter((e) => e.impact === "under review").length;
  const totalEvolutions = files.reduce((s, f) => s + f.evolution, 0);

  return { files, totalEvolutions, lastReflection, pendingCount };
}

export function getPendingEvolutions(): MemoryEntry[] {
  const memFile = readFile("memories.md");
  const entries = loadMemoryEntries(memFile.raw);
  return entries.filter((e) => e.impact === "under review");
}

File: server-py/src/ai_personality/personality.py
import os
import re
from pathlib import Path

import yaml

from .defaults import generate_default_files

FILE_NAMES = [
    "identity.md",
    "traits.md",
    "values.md",
    "rules.md",
    "memories.md",
    "relationships.md",
]


def get_personality_dir() -> Path:
    env = os.environ.get("AI_PERSONALITY_DIR")
    if env:
        return Path(env).resolve()
    return Path.home() / ".ai-personality" / "personality"


PERSONALITY_DIR = get_personality_dir()
_init_done = False


def ensure_personality_dir() -> None:
    global _init_done
    if _init_done:
        return
    _init_done = True
    PERSONALITY_DIR.mkdir(parents=True, exist_ok=True)
    print(
        f"Created personality directory: {PERSONALITY_DIR}",
        file=__import__("sys").stderr,
    )
    for filename, content in generate_default_files().items():
        fp = PERSONALITY_DIR / filename
        if not fp.exists():
            fp.write_text(content, encoding="utf-8")
            print(f"Created {filename}", file=__import__("sys").stderr)


def _parse_frontmatter(raw: str) -> tuple[dict | None, str]:
    m = re.match(r"^---\n(.*?)\n---\n(.*)", raw, re.DOTALL)
    if not m:
        return None, raw
    data = yaml.safe_load(m.group(1))
    return data or {}, m.group(2).strip()


def _read_file(filename: str) -> dict:
    ensure_personality_dir()
    fp = PERSONALITY_DIR / filename
    raw = fp.read_text("utf-8")
    fm, body = _parse_frontmatter(raw)
    return {
        "filename": filename,
        "frontmatter": fm or {},
        "body": body,
        "raw": raw,
    }


def _load_memory_entries(raw: str) -> list[dict]:
    marker = "<!-- entries -->"
    idx = raw.find(marker)
    if idx == -1:
        return []
    after = raw[idx + len(marker) :]
    m = re.search(r"```yaml\n(.*?)```", after, re.DOTALL)
    if not m:
        return []
    entries = yaml.safe_load(m.group(1))
    if not entries:
        return []
    return entries if isinstance(entries, list) else [entries]


def _dump_yaml(data) -> str:
    out = yaml.dump(data, default_flow_style=False, allow_unicode=True, sort_keys=False)
    out = re.sub(r"['\"](\d{4}-\d{2}-\d{2})['\"]", r"\1", out)
    return out


def read_personality_file(filename: str) -> dict:
    if filename not in FILE_NAMES:
        raise ValueError(f"Unknown personality file: {filename}")
    return _read_file(filename)


def read_all_personality_files() -> list[dict]:
    return [_read_file(f) for f in FILE_NAMES]


def get_summary() -> str:
    parts = []
    for f in read_all_personality_files():
        name = f["filename"].replace(".md", "")
        fm = f["frontmatter"]
        preview = f["body"][:300]
        if len(f["body"]) > 300:
            preview += "..."
        parts.append(
            f"=== {name} ({fm.get('type', '?')}, evolution {fm.get('evolution', 0)}, {len(fm.get('crossReferences', []))} refs) ===\n{preview}"
        )
    return "\n\n".join(parts)


def reflect(
    experience: str, lesson: str, affected_files: list[str] | None = None
) -> dict:
    from datetime import date

    fp = PERSONALITY_DIR / "memories.md"
    raw = fp.read_text("utf-8")
    today_str = date.today().isoformat()
    new_entry = {
        "date": today_str,
        "experience": experience,
        "lesson": lesson,
        "impact": "under review",
        "affectedFiles": affected_files or [],
    }
    marker = "<!-- entries -->"
    idx = raw.find(marker)
    if idx == -1:
        raw += f"\n\n{marker}\n\n```yaml\n{_dump_yaml([new_entry])}```\n"
    else:
        after = raw[idx + len(marker) :]
        ym = re.search(r"```yaml\n(.*?)```", after, re.DOTALL)
        if ym:
            abs_start = idx + len(marker) + ym.start()
            abs_end = abs_start + len(ym.group(0))
            existing = yaml.safe_load(ym.group(1)) or []
            if not isinstance(existing, list):
                existing = [existing]
            existing.append(new_entry)
            raw = (
                raw[:abs_start]
                + "```yaml\n"
                + _dump_yaml(existing)
                + "```"
                + raw[abs_end:]
            )
        else:
            insert = idx + len(marker)
            raw = (
                raw[:insert]
                + f"\n\n```yaml\n{_dump_yaml([new_entry])}```\n"
                + raw[insert:]
            )

    fp.write_text(raw, "utf-8")
    entries = _load_memory_entries(raw)
    return {"date": today_str, "entries": entries}


def get_status() -> dict:
    files = []
    total_evo = 0
    for f in read_all_personality_files():
        fm = f["frontmatter"]
        files.append(
            {
                "name": f["filename"],
                "evolution": fm.get("evolution", 0),
                "lastUpdated": str(fm.get("lastUpdated", "?")),
                "refCount": len(fm.get("crossReferences", [])),
            }
        )
        total_evo += fm.get("evolution", 0)

    mem = _read_file("memories.md")
    entries = _load_memory_entries(mem["raw"])
    last_ref = entries[-1]["date"] if entries else None
    pending = sum(1 for e in entries if e.get("impact") == "under review")

    return {
        "files": files,
        "totalEvolutions": total_evo,
        "lastReflection": last_ref,
        "pendingCount": pending,
    }


def get_pending_evolutions() -> list[dict]:
    mem = _read_file("memories.md")
    entries = _load_memory_entries(mem["raw"])
    return [e for e in entries if e.get("impact") == "under review"]
